### ISSUE TITLE
Potential issue with keras version 2.2.4 failing to train

### ISSUE THREAD

**Issue by colinsauze** [carpentries-incubator/deep-learning-intro#102]
Two learners during the trial lesson on june 22nd 2021 reported problems with loss rates not dropping and the network not converging on a solution. Both seem to have used a fresh conda environment and ended up with keras version 2.2.4, yet I seem to have 2.4.0. I have a copy of their notebooks and conda environments to investigate the causes further.

**Comment by psteinb** [comment]
It might be worth considering, to put the model code into unit tests and run them through github actions. This way we could persistify this behavior.

**Comment by dafnevk** [comment]
This is odd. Could you indeed share their conda environments? I'll try to see if I can reproduce their problem. Of course, we could require a higher tensorflow / keras version in the setup instructions but then I'm not sure which version this should be.

**Comment by dafnevk** [comment]
Tensorflow 2.2 has keras version 2.3 (quite confusing...) so maybe we can set that as requirement.

**Comment by colinsauze** [comment]
> It might be worth considering, to put the model code into unit tests and run them through github actions. This way we could persistify this behavior.

I like that idea, but is there an easy way to synchronise an external .py file with the markdown (or extract the code from the markdown)?

We have quite a few places where we present a few code fragments separately which together make a whole program. We would need to combine all of the appropriate fragments into one python script in order to test it.

**Comment by colinsauze** [comment]
Here are the environment files and notebooks used by both of the learners who had this problem. 

[problem_envs.zip](https://github.com/carpentries-incubator/deep-learning-intro/files/6711106/problem_envs.zip)

**Comment by florian-huber** [comment]
keras `2.2.4` is one of the last versions to work with Tensorflow 1, so although it sounds like it would be close to `2.3` or `2.4` I think it is not. 
--> https://github.com/keras-team/keras/releases

We could also check if some of the key defaults changed (e.g. learning rate).

**Comment by bpmweel** [comment]
Another easy solution during teaching is to have the people that are on keras 2.2.4 or lower to use the `values` of the pandas dataframe at training. So as follows:

`history = model.fit(X_train.values, y_train.values, epochs=100)`

**Comment by psteinb** [comment]
Returning to the question, @colinsauze brought up: I think we are technology-wise in a tricky or challenging situation.

- we could outsource code snippets (which we consume inside the markdowns) and only `include` them inside the jekyll processed markdown files. this can provide a basis for the unit tests, but makes the markdown not readable

- the carpentries jekyll infrastructure provides the facility for us to use Rmarkdown. If we choose to do so, we might be able to use the [reticulate](https://rstudio.github.io/reticulate/) package and other libraries to make python code run within `.Rmd` files. 

- we use nbconvert to run dedicated jupyter notebooks: https://nbconvert.readthedocs.io/en/latest/execute_api.html

Any feedback at this point? If not I suggest to post a question on the lesson-dev slack channel.

**Comment by bpmweel** [comment]
@psteinb I have made (or blatently stole and adapted) a [tool](https://github.com/bpmweel/cce) that extracts all the code snippets from a markdown file, saves them to a python file and runs the python file. I also used this in a github action in this pull request: https://github.com/carpentries-incubator/deep-learning-intro/pull/108

The tool assumes the following:
- All code snippets continue from the last one
- They run without error

Some things that still may need improvements:
- There is no checking of the output

**Comment by colinsauze** [comment]
If we use nbconvert would that require all code to be inside the notebooks instead of the lessons?

**Comment by dafnevk** [comment]
Can't we require a higher version of Keras to avoid issues? PR #145 proposes to require tensorflow >= 2.5 (which also means Keras 2.5. It seems that starting at v2.4, Keras is completely inside tensorflow and the versions match).

**Comment by dafnevk** [comment]
Resolved with PR #145
